Call proxy attribute methods for reserved aliases

nvasilevski opened the pull request:
This PR allows aliasing columns named with one of the reserved
methods like `id`, `id_before_type_cast`, `id_was`, etc if database
schema happed to have one of these.

Currently there is no convenient way of reading the `id` column value
which is not being used as a primary key as `id` in Rails is a reserved
term which means "an identifier" or the value stored in the primary key
column.

Given a `posts` table with both `id` and `title` columns where `title`
is being used as a primary key it is possible to alias `id` attribute
to a new name so we can fetch the column value.

For example:
```ruby
class Post < ActiveRecord::Base
  self.primary_key = :title
  alias_attribute :post_id, :id
end

post = Post.create(title: "Hello world")
post.post_id # => 1
```

### Technical details

Currently aliasing an attribute like `alias_attribute :subject, :title` will define `<prefix_subject_suffix>` methods that directly delegate to the `*title*` methods. For example the `subject` reader would look like:

```ruby
def subject
  title
end
```
This works fine but doesn't work when aliasing any of the Active Record reserved methods so in current implementation an alias like `alias_attribute :subject, :id` ends up generating

```ruby
def subject
  id
end
```

which will return the value behind the `primary_key` column and not the `id` column value. 

With the change we are proposing aliasing reserved method will generate methods that call into so-called "proxy" methods that read directly from the attributes. So aliasing `id` attribute will generate:

```ruby
def subject
  attribute("id")
end
```
https://github.com/rails/rails/blob/d503e893056650524caaf9ffe0364417224c051a/activemodel/lib/active_model/attributes.rb#L148-L149

along with `attribute_was?` `attribute_changed?`, etc and allow to access the `id` value directly. 

Since now we have versions of the alias attribute method body we had to make sure we are using two different namespaces to avoid one of the bodies being cached. This is to prevent issues when a non Active Record model, like an Active Model model aliases `id` and pollutes the `:alias_attribute` namespace cache with a non-primary key implementation. 

### Long-term plans

Despite of the solution being generic enough to allow for any of the reserved method names to be aliased, Rails currently disallows having schema with columns named as one of the reserved methods. The exception will be raised from:
https://github.com/rails/rails/blob/e7f7a13950abfd980089fc581b56128ea4a04f8e/activerecord/lib/active_record/attribute_methods.rb#L89
However eventually Rails may end up relaxing this restriction assuming that a model could alias the reserved method column. For example if the application ends up in a situation where it needs to work with a schema that contains columns named with reserved terms it could alias these columns like:

```ruby
alias_attribute :delete_value, :delete
alias_attribute :save_value, :save
alias_attribute :update_value, :update
...
```

### Side note

Overall it looks like the `alias_attribute` generation could be simplified and we could at least get rid of one of the regexes 
https://github.com/rails/rails/blob/d503e893056650524caaf9ffe0364417224c051a/activemodel/lib/active_model/attribute_methods.rb#L67-L68

Also `COMPILABLE` term doesn't seem to be the best fit for the logic anymore. As far as I can tell it used to differentiate between methods built inline like a string (compiled) and methods that were defined using `define_method` 
We are not using `define_method` anymore which means both of the branches are "complied" in the legacy terms.
So it looks like we can refactor the logic to only do the `.match` once to determine whether we want a `send` body or `self.method` body. It may not result in any performance gains but overall the logic is complex and it would be nice to simplify it a bit.

nvasilevski committed:
Call proxy attribute methods for reserved aliases

This commit allows aliasing columns named with one of the reserved
methods like `id`, `id_before_type_cast`, `id_was`, etc if database
schema happed to have one of these.

Currently there is no convenient way of reading the `id` column value
which is not being used as a primary key as `id` in Rails is a reserved
term which means "an identifier" or the value stored in the primary key
column.

Given a `posts` table with both `id` and `title` columns where `title`
is being used as a primary key it is possible to alias `id` attribute
to a new name so we can fetch the column value.

For example:
```ruby
class Post < ActiveRecord::Base
  self.primary_key = :title
  alias_attribute :post_id, :id
end

post = Post.create(title: "Hello world")
post.post_id # => 1
```

Co-authored-by: Paarth Madan <paarth.madan@shopify.com>

nvasilevski commented on activerecord/lib/active_record/attribute_methods.rb:
```diff
@@ -106,6 +106,11 @@ def dangerous_attribute_method?(name) # :nodoc:
       end
 
+      # Same as dangerous_attribute_method? but includes primary key methods.
+      def framework_reserved_method?(method_name) # :nodoc:
```
We can't use `dangerous_attribute_method?` as it excludes primary key methods from the list and we want them to be treated as "dangerous" for the purpose of aliasing.

https://github.com/rails/rails/blob/a5a19ff5a5278081e5c7b251d68cf39300555d7f/activerecord/lib/active_record/attribute_methods/primary_key.rb#L98

eileencodes commented on activemodel/lib/active_model/attribute_methods.rb:
```diff
@@ -234,6 +243,34 @@ def alias_attribute(new_name, old_name)
         end
       end
 
+      def compilable_alias_attribute_method_body_for(old_name, pattern, parameters) # :nodoc:
```
Rather than no-doc these, they can be private since they aren't called outside this file.

eileencodes commented on activerecord/test/cases/attribute_methods_test.rb:
```diff
@@ -30,6 +31,24 @@ def setup
     ActiveRecord::Base.send(:attribute_method_patterns).concat(@old_matchers)
   end
 
+  test "aliasing `id` attribute allows reading the column value" do
```
I think we need more tests in active model, otherwise this behavior is untested over there and anyone using active model without active record isn't covered by the tests.

nvasilevski replied:
The problem with testing it directly in Active Model is that we can't come up with a reasonable scenario that represents a real application use-case. 
By having `framework_reserved_method?` method explicitly returning `false` we essentially say "nothing is going to change for Active Model" and in order to verify that the only thing we can do is to run existing tests.

It would be important to add Active Model tests when `framework_reserved_method?` for Active Model becomes smarter but for now we can't test much unless we stub `framework_reserved_method?` in the test but that won't be testing a realistic scenario

eileencodes reviewed (changes requested):
This also needs a CHANGELOG entry. 

@matthewd can you review again?

nvasilevski commented:
Rebased, added a changelog entry, moved new methods to `private` along with adding a tiny bit of internal documentation since it's not very easy to guess what exactly is being generated by these methods. 

I haven't added tests to Active Model since the intention currently is to only affect an `ActiveRecord::Base` models while active model models will keep using `compilable_alias_attribute_method_body_for` and `dynamic_alias_attribute_method_body_for` methods which stay unchanged

eileencodes commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,25 @@
+*   `alias_attribute` can be used to alias columns named with a primary key reserved terms.
```
It's not just primary keys though right? We're improving alias attribute for any dangerous attribute method so I think we should mention that.

nvasilevski replied:
Technically yes, we made it possible to use `alias_attribute` with more than just pk-related terms, for example:

```ruby
alias_attribute :get_update_value, :update
alias_attribute :get_save_value, :save
alias_attribute :get_destroy_value, :destroy
```

But it's tricky because Rails won't let you to boot an app with columns named `update`, `destroy` or `update`. 
So it will be possible to use this feature without changes to the `alias_attribute` logic if Rails loosens the following restriction:
https://github.com/rails/rails/blob/2e181b1b2edbeed0c6afb11cc037dd23c985fdb9/activerecord/lib/active_record/attribute_methods.rb#L89

I can change the description to cover all reserved methods as it is correct from technical standpoint but it won't be fair as there will be no applications with columns named with most of reserved terms

matthewd commented on activemodel/lib/active_model/attribute_methods.rb:
```diff
@@ -417,6 +430,42 @@ def define_proxy_call(code_generator, name, proxy_target, parameters, *call_args
           end
         end
 
+        # Generates body that calls the original method directly.
```
Elsewhere I've just recommended that we avoid some of these constructs entirely by calling existing methods... for those that remain (or even if we can't do that, and have to keep all of the below behaviour), I think they're short enough (and no longer overridden anywhere, which I think might've been the original motivation for extraction?) that it's better for readability to keep this all in-line in the calling method.
